### ISSUE TITLE
[Proposal] Add an additional API to dapr for getting metadata on a workflow component

### ISSUE THREAD

**Issue by RyanLettieri** [dapr/dapr#7025]
## In what area(s)?

/area runtime

> /area operator

> /area placement

> /area docs

> /area test-and-release
## Describe the proposal
This proposal stems from the following issue: https://github.com/dapr/dapr/issues/6402 
Keeping track of all your created workflows can be tedious, especially when working on a large scale. With this newly proposed API for getting workflow metadata, users will now be able to access and obtain more information on their workflows from a simple API call. The metadata items provided from this call are as follows:

- Names of registered workflows
- Names of registered activities associated for each individual workflow type

The proposed API is as follows:
`workflows/{workflowComponent}/{getMetadata}/?workflowName={workflowName}`

- workflows: The start of all workflow APIs
- workflowComponent: The registered workflow component such as 'Dapr'
- getMetadata: The name of the proposed API
- workflowName: The ***optional*** parameter that when specified, will limit the metadata returned to only information pertaining to that individual workflow

### Example of a returned metadata structure
```
  "workflows":[
       {
           "name":"workflowName1"
            {
               "registeredActivites:
                  {
                       "activityA",
                       "activityB",
                       "activityC",
                  }
            },
           "name":"workflowName2"
            {
               "registeredActivites:
                  {
                       "activityD",
                       "activityE",
                       "activityF",
                  }
            },
       }
   ],
```

### Additional Information and Concerns

- While it would be nice to include even more information in these structures such as instance IDs for both completed and active workflow instances, the amount of data returned from this API could end up being very large.
- Due to the nature in which the apps+dapr will scale out, it is possible that a workflow instance is started on one node and completes on another. When querying for metadata, the node in which the workflow started on will say that the workflow is still running, whereas the node in which the workflow completed on will say that the workflow has completed. This inconsistency between the nodes leads to the inability to get the number of running + completed workflow instances across all nodes.

### Counter Proposal
Another user has created an issue/proposal with the intention of adding workflow metadata directly to the current metadata API. This proposal can be viewed here:  https://github.com/dapr/dapr/issues/6402 

While it is a reasonable solution and can be handled by only providing this information if a workflow component is registered, it may make more sense to have this metadata available through an additional API as referenced above in this proposal.

**Comment by ItalyPaleAle** [comment]
I think this is a good idea, but replace the endpoint with a more REST-ful one:

```
workflows/{workflowComponent}/getMetadata/{workflowName}
```

Also, what would the gRPC endpoint be?

**Comment by olitomlinson** [comment]
My 2c 

From an end-user perspective, having a single `metadata` endpoint is a better experience. It's easier to remember just one endpoint i.e. `v1.0/metadata` over having to remember several different endpoints with different conventions etc

So I would prefer that the existing `v1.0/metadata` endpoint is extended, to include what you propose.

- Added a `dapr` object under workflows, this then allows the model to be extended to other 3rd party workflow systems in future, that may have different properties.
- Modified the `Activities` into an `object`, so additional data could be appended in future if there was a need.

```json
{
  "id": "demo-actor",
  "actors": [
    {
      "type": "DemoActor",
      "count": 1
    }
  ],
  "extended": {
    "myDemoAttribute": "myDemoAttributeValue",
    "cliPID": "1031040",
    "appCommand": "uvicorn --port 3000 demo_actor_service:app"
  },
  "components": [
    {
      "name": "pubsub",
      "type": "pubsub.redis",
      "version": "v1",
      "capabilities": [
        ""
      ]
    }
  ],
  "workflows": {
    "dapr": [
      {
        "name": "workflowName1",
        "registeredActivites": [
          {
            "name": "activityA"
          },
          {
            "name": "activityB"
          },
          {
            "name": "activityC"
          }
        ]
      }
    ]
  }
}
```

My assumption is also that given that the request would be targeting a `metadata` endpoint on a specific sidecar, this sidecar would only have awareness of the workflows and activities that are registered to this particular `app` (it wouldn't have a global awareness of workflows registered to other apps) - I'm ok with this lack of global awareness, as workflows are scoped to individual apps anyway.

**Comment by olitomlinson** [comment]
> - While it would be nice to include even more information in these structures such as instance IDs for both completed and active workflow instances, the amount of data returned from this API could end up being very large.
> - Due to the nature in which the apps+dapr will scale out, it is possible that a workflow instance is started on one node and completes on another. When querying for metadata, the node in which the workflow started on will say that the workflow is still running, whereas the node in which the workflow completed on will say that the workflow has completed. This inconsistency between the nodes leads to the inability to get the number of running + completed workflow instances across all nodes.

I agree it would be a significant challenge to maintain a centralised aggregate of the various states of all workflows. I believe this should be passed on to the end-user to implement via a firehose of lifecycle events (this is what Azure Durable Functions do)

I have an existing proposal for this capability https://github.com/dapr/dapr/issues/6072

**Comment by cgillum** [comment]
Regarding `registeredActivites`, the current programming model doesn't require workflows to declare what their activities are. Rather, activities are declared at the app level, and the activities that actually get called by any given workflow is dynamic and unknowable to a large degree.

To make it possible to display activities under a workflow node, we would need to change the programming model for all workflow SDKs to have workflows explicitly register their activity dependencies.

**Comment by olitomlinson** [comment]
hmmm... What would be the value in explicitly registering Activities with a Workflow, other than for reporting purposes? I can't see any at this point? So that would seem a significant change, with little value.

If reporting _is_ the value, I would expect the Observability data to be able to tell that story...

Anyway, considering your feedback Chris, the json could be restructured as such :

```json
"workflows": {
    "dapr": {
      "workflows": [
        {
          "name": "workflowName1"
        }
      ],
      "activites": [
        {
          "name": "activityA"
        },
        {
          "name": "activityB"
        },
        {
          "name": "activityC"
        }
      ]
    }
  }
```

**Comment by olitomlinson** [comment]
I would like to see this implemented.

I utilise the `/metadata` endpoint all the time when trying to understand the sidecars view of the world compared to what I think the sidecars view of the world should look like.

And so, for the same reason, I see extending the `/workflow` endpoint to include workflows would be really advantageous from a self-serve debugging/diagnostics perspective.

**Comment by olitomlinson** [comment]
Keep-alive

**Comment by olitomlinson** [comment]
Bump
